spec/src/main/asciidoc/repository.asciidoc

```diff
@@ -112,6 +112,246 @@ Subsequently, an implementation of Jakarta Data, specifically tailored to the ch
+- be of type `Limit`, `Pageable`, or `Sort`.
+
+A repository with annotated query methods with named parameters should be compiled so that parameter names are preserved in the class file (for example, using `javac -parameters`), or the parameter names must be specified explicitly using the `@Param` annotation.
+
+A parameter of an annotated query method may not be annotated with lifecycle annotation.
+
+The return type of the annotated query method must be consistent with the result type of the query specified by the query annotation.
+
+[NOTE]
+====
+The result type of a query depends on datastore-native semantics, and so the return type of an annotated query method cannot be specified here.
+However, Jakarta Data implementations are strongly encouraged to support the following return types:
+
+- for a query which returns a single result of type `T`, the type `T` itself, or `Optional<T>`,
+- for a query which returns many results of type `T`, the types `List<T>`, `Page<T>`, and `T[]`.
+
+Furthermore, implementations are encouraged to support `void` as the return type for a query which never returns a result.
+====
+
+This specification defines the built-in `@Query` annotation, which may be used to specify a query in an arbitrary query language understood by the Jakarta Data provider.
+
+For example, using a named parameter:
+
+[source,java]
+----
+@Query(“where title like :title order by title”)
+Page<Book> booksByTitle(String title, Pageable page);
+----
+
+Or, using a positional parameter:
+
+[source,java]
+----
+@Query(“delete from Book where isbn = ?1”)
+void deleteBook(String isbn);
+----
+
+Programs which make use of annotated query methods are not portable between providers.
+
+[NOTE]
+====
+A Jakarta Data provider might extend this specification to define its own query annotation types.
+For example, a provider might define a `@SQL` annotation for declaring queries written in SQL.
+====
+
+There is no special programming model for query annotations.
+The Jakarta Data implementation automatically recognizes the query annotations it supports.
+
+==== Automatic query methods
+
+An _automatic query method_ is an abstract method with return type `E`, `Optional<E>`, `Page<E>`, or `List<E>`, where `E` is an entity class, and where each parameter either:
+
+- has exactly the same type and name as a persistent field or property of the entity class, or
+- is of type `Limit`, `Pageable`, or `Sort`.
+
+A repository with automatic query methods should be compiled so that parameter names are preserved in the class file (for example, using `javac -parameters`), or the parameter names must be specified explicitly using the `@Param` annotation.
+
+A parameter of an automatic query method may not be annotated with lifecycle annotation.
+
+For example:
+
+[source,java]
+----
+Book bookByIsbn(String isbn);
+
+List<Book> booksByYear(Year year, Sort order, Limit limit)
```

Comment:
When I previously saw "Automatic Query methods" in the list of specification-defined types of repository methods,

> - an entity instance _lifecycle method_,
> - an _annotated query method_,
> - an _automatic query method_, or
> - a _resource accessor method_.

I assumed that the term "automatic query method" encompassed the Query by Method Name pattern as well as the pattern described in this section because both automatically generate a query.  This section either needs a more specific name and the previously mentioned list needs to be updated, or we need to update this section to include Query by Method Name.  I'll explore the latter with a suggestion,

```suggestion
==== Automatic query methods

An _automatic query method_ is an abstract method that either follows the Query by Method Name pattern where the Jakarta Data provider generates a query based on the name of the method, or follows a pattern for whereby the Jakarta Data provider automatically generates a find query based on the names of parameters that are supplied to the method, where the method has return type that identifies the entity, such as `E`, `Optional<E>`, `Page<E>`, or `List<E>`, where `E` is an entity class. Each parameter must either:

- have exactly the same type and name as a persistent field or property of the entity class, or
- be of type `Limit`, `Pageable`, or `Sort`.

A repository with automatic query methods that are based on parameters must either be compiled so that parameter names are preserved in the class file (for example, using `javac -parameters`), or the parameter names must be specified explicitly using the `@Param` annotation.

For example:

[source,java]
----
Book bookByIsbn(String isbn);

List<Book> booksByYear(Year year, Sort order, Limit limit)
```

Reply:
> I assumed that the term "automatic query method" encompassed the Query by Method Name pattern as well as the pattern described in this section because both automatically generate a query. 

It could, I still have not thought this through completely.